### ISSUE TITLE
Vim9: function optional string argument gets perceived as 'any' when defaulting

### ISSUE THREAD

**Issue by mark2185** [vim/vim#9376]
### Steps to reproduce

Run this shell command
```
vim -Nu NONE -S <(cat <<'EOF'
    vim9script
    g:variable = get( g:, 'variable', '' )
    if 1 == type(g:variable) | echom "I am a string" | endif
    def Func(arg: string = g:variable ): void
    enddef
    Func()
EOF
)
```

This is the output:
```
I am a string
Error detected while compiling command line..script /proc/42203/fd/11[6]..function <SNR>1_Func:
E1013: Argument 1: type mismatch, expected string but got any
Press ENTER or type command to continue
```

### Expected behaviour

It echoes `I am a string` and throws no errors.

### Version of Vim

8.2.1-3582

### Environment

It's on `archlinux`, with `terminator` as terminal.

**Comment by brammool** [comment]
The error is correct.  The default value is not computed when the function is defined, but turned into an expression.
A global variable can get any value later, thus the type is "any", which does not satisfy the required type.
You can add a type cast, which will result in checking the type at runtime.  You can see this using ":disassemble":

```
def Func(arg: string = <string>g:variable ): void
```
  arg = <string>g:variable
   0 JUMP_IF_ARG_SET arg[-1] -> 4
   1 LOADG g:variable
   2 CHECKTYPE string stack[-1]
   3 STORE arg[-1]

```

```

**Comment by lacygoill** [comment]
It seems inconsistent for an error to be given when `arg` is a function argument:
```vim
vim9script
g:variable = 'string'
def Func(arg: string = g:variable)
enddef
defcompile
```
    E1013: Argument 1: type mismatch, expected string but got any

But not when `arg` is a function variable:
```vim
vim9script
g:variable = 'string'
def Func()
    var arg: string = g:variable
enddef
defcompile
```
    no error

**Comment by lacygoill** [comment]
Also, it is not obvious to understand how useful type cast is from the example given at [`:help type-casting`](https://vimhelp.org/vim9.txt.html#type-casting):

    var l: list<number> = [1, g:two]

Because it doesn't give any error.

Neither at the script level:
```vim
vim9script
g:two = 2
var l: list<number> = [1, g:two]
```
    no error

Nor in compiled code:
```vim
vim9script
g:two = 2
def Func()
    var l: list<number> = [1, g:two]
enddef
Func()
```
    no error

After executing `:disassemble`, we can see that a type cast does make a difference:
```diff
--- /tmp/after	2021-12-20 20:36:38.717248117 +0100
+++ /tmp/before	2021-12-20 20:36:20.261308482 +0100
@@ -1,9 +1,9 @@
 <SNR>1_Func
-    var l: list<number> = [1, g:two]
+    var l: list<number> = [1, <number>g:two]
    0 PUSHNR 1
    1 LOADG g:two
-   2 NEWLIST size 2
-   3 CHECKTYPE list<number> stack[-1]
+   2 CHECKTYPE number stack[-1]
+   3 NEWLIST size 2
    4 SETTYPE list<number>
    5 STORE $0
    6 RETURN void
```
But I don't think the reader will think about executing `:disassemble`.  And they should not be expected to understand its output, since the generated instructions like `CHECKTYPE` and `NEWLIST` are not documented.

To illustrate how useful a type cast is, the example given in the OP seems more useful.

---

Also, this sentence is weird:

   > The compiled code will then only check that "g:two" is a number and give an
   > error if it isn't.

The purpose of a type cast is to delay type checking until runtime, right?  So why does this sentence mention "compiled code", and why doesn't it mention "runtime"?

---

Anyway, as a suggestion, here is a patch which tries to improve the help:
```diff
diff --git a/runtime/doc/vim9.txt b/runtime/doc/vim9.txt
index 39affbfc2..7af012a5f 100644
--- a/runtime/doc/vim9.txt
+++ b/runtime/doc/vim9.txt
@@ -1277,16 +1277,23 @@ value can be changed at any time, possibly changing the type.  Therefore, in
 compiled code the "any" type is assumed.
 
 This can be a problem when the "any" type is undesired and the actual type is
-expected to always be the same.  For example, when declaring a list: >
-	var l: list<number> = [1, g:two]
-At compile time Vim doesn't know the type of "g:two" and the expression type
-becomes list<any>.  An instruction is generated to check the list type before
-doing the assignment, which is a bit inefficient.
+expected to always be the same: >
+	vim9script
+	g:global = 'string'
+	def Func(arg: string = g:global)
+	enddef
+	defcompile
+At compile time Vim doesn't know the type of "g:global", and assumes "any"
+which doesn't match the "string" type in the "arg" declaration.
 							*type-casting*
 To avoid this, use a type cast: >
-	var l: list<number> = [1, <number>g:two]
-The compiled code will then only check that "g:two" is a number and give an
-error if it isn't.  This is called type casting.
+	vim9script
+	g:global = 'string'
+	def Func(arg: string = <string>g:global)
+	enddef
+	defcompile
+Type checking will then only check that "g:global" is a string at runtime, and
+give an error if it isn't.  This is called type casting.
 
 The syntax of a type cast is:  "<" {type} ">".  There cannot be white space
 after the "<" or before the ">" (to avoid them being confused with
```

**Comment by brammool** [comment]
> It seems inconsistent for an error to be given when `arg` is a function argument:
> ```vim
> vim9script
> g:variable = 'string'
> def Func(arg: string = g:variable)
> enddef
> defcompile
> ```
>     E1013: Argument 1: type mismatch, expected string but got any
> 
> But not when `arg` is a function variable:
> ```vim
> vim9script
> g:variable = 'string'
> def Func()
>     var arg: string = g:variable
> enddef
> defcompile
> ```
>     no error

It is hard to decide where to make the type checks happen at compile
time and where it is OK to postpone until runtime.

The advantage of a compile time check is that an error can be given very
early, before the script is even executed.  That makes it a lot easier
to write correct code.

The problem with using global variables is that their type can change
any moment.  They can even be deleted.  So what will we do with them?

The check for the variable not existing obviously needs to be done at
runtime.  Thus a runtime check is always needed.  My previous suggestion
that a type cast should be added doesn't make sense in that context.

Let's compare with a script variable, for which we can define the type:

	var s: string
	def Func(arg: string = s)

Obviously correct.

	var nr: number
	def Func(arg: string = nr)

Obviously an error.

	var a: any
	def Func(arg: string = a)

Now this is a place where we can make a choice.  It might be that the
user accidentally used the wrong variable.  Requiring a type cast makes
sense here, it makes clear that the type might be wrong and we know that:

	var a: any
	def Func(arg: string = <string>a)

For global variables we never know the type.  Thus the reasoning that
applies to local variables would not apply.  Is that enough to make a
different choice?

It also should be easy to explain.  If we say "when using a global
variable there always is a runtime type check" is simple.  So I'm
thinking we should do it that way.

Note that the same applies to buffer-local, window-local and legacy
script variables.

-- 
From "know your smileys":
 ~#:-(	I just washed my hair, and I can't do nuthin' with it.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
There seems to be another inconsistency.  This time, for local variables:
```vim
vim9script
var a: any
def Func(arg: string = a)
enddef
defcompile
```
    E1013: Argument 1: type mismatch, expected string but got any
```vim
vim9script
var a: any
def Func()
    var arg: string = a
enddef
defcompile
```
    no error

Why does assigning `a` to a string variable gives an error in the function's header, but not in its body?

To be consistent, I think the second snippet should give an error, just like the first one.

Note that, when we declare a variable with the type `any`, by default it gets assigned the number 0:
```vim
vim9script
var a: any
echo a
```
    0

Now, watch what happens to the previous snippet – the one which unexpectedly doesn't give any error – if we replace `: any` with `= 0`:
```vim
vim9script
var a = 0
def Func()
    var arg: string = a
enddef
defcompile
```
    E1012: Type mismatch; expected string but got number

This time, an error is correctly given.

**Comment by brammool** [comment]
> There seems to be another inconsistency.  This time, for local variables:
> ```vim
> vim9script
> var a: any
> def Func(arg: string = a)
> enddef
> defcompile
> ```
>     E1013: Argument 1: type mismatch, expected string but got any
> ```vim
> vim9script
> var a: any
> def Func()
>     var arg: string = a
> enddef
> defcompile
> ```
>     no error
> 
> Why does assigning `a` to a string variable gives an error in the function's header, but not in its body?
> 
> To be consistent, I think the second snippet should give an error, just like the first one.
> 
> Note that, when we declare a variable with the type `any`, by default it gets assigned the number 0:
> ```vim
> vim9script
> var a: any
> echo a
> ```
>     0
> 
> Now, watch what happens to the previous snippet – the one which unexpectedly doesn't give any error – if we replace `: any` with `= 0`:
> ```vim
> vim9script
> var a = 0
> def Func()
>     var arg: string = a
> enddef
> defcompile
> ```
>     E1012: Type mismatch; expected string but got number
> 
> This time, an error is correctly given.

Yes, if we consequently disallow assigning "any" to a specific type,
then we have to make more changes.

I started doing this, but it turns out quite a few things start failing.
For example:

	def Func(arg: any): string
	   return arg
	enddef

That now needs to be:

	def Func(arg: any): string
	   return <string>arg
	enddef

It gets more complicated when the argument can be different types.  This
actually happens in the test functions, where the "error" argument can
be a string or a list of strings:

	def CheckDefAndScriptFailure(lines: list<string>, error: any, lnum = -3)
	  var errorDef: string
	  var errorScript: string
	  if type(error) == v:t_string
	    errorDef = <string>error
	    errorScript = <string>error
	  elseif type(error) == v:t_list && len(error) == 2
	    errorDef = (<list<string>>error)[0]
	    errorScript = (<list<string>>error)[1]
	  else

It does seem correct, but we'll need type casts in more places.

-- 
From "know your smileys":
 |-P	Reaction to unusually ugly C code

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> I started doing this, but it turns out quite a few things start failing.

> It does seem correct, but we'll need type casts in more places.

I don't know any other programming language, but type casts look a bit weird to me.  I would naively think "the fewer we need them, the better".

If this change breaks a lot of existing tests, maybe it's a hint that it's not a good idea.  I don't know.

**Comment by brammool** [comment]
> > I started doing this, but it turns out quite a few things start failing.
> 
> > It does seem correct, but we'll need type casts in more places.
> 
> I don't know any other programming language, but type casts look a bit
> weird to me.  I would naively think "the fewer we need them, the
> better".
> 
> If this change breaks a lot of existing tests, maybe it's a hint that
> it's not a good idea.  I don't know.

I'm in doubt.  In some places it makes sense, but in some places,
especially when using builtin functions that may return one type or
another, it gets annoying.  With list-unpack assignment it's even not
possible, thus I have to make an exception for that.

But the tests tend to cover corner cases, in a normal script there would
be far fewer need for type cases.

-- 
From "know your smileys":
 :q	vi user saying, "How do I get out of this damn emacs editor?"

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
When starting a script, I'm often giving the `any` type to a lot of variables/arguments, because I don't want to be bothered by type checking yet.  It's only after the code compiles that I start replacing `any` with more specific types.  I suspect other people do (or will do) that too.  If we give this error, I'm concerned that this development phase will become much more painful.

FWIW, instead of making the second snippet give an error, I would suppress the error in the first one, so that both of them are consistent, without annoying the user:
```vim
vim9script
var a: any
def Func(arg: string = a)
enddef
defcompile
```
    # now: E1013: Argument 1: type mismatch, expected string but got any
    # in the future: no error

Maybe we could add an item at the end of the Vim9 todo list, to think about it more.  The language has been really solid the last few months.  No major breaking change which require large refactorings.

> With list-unpack assignment it's even not possible, thus I have to make an exception for that.

It looks like another hint that it's a bad idea.  Also, currently, for a given variable/argument, we only have to give the type once; in the declaration.  But with a type cast, we might need to give it several times; possibly as many times as we refer to the variable.  That seems too cumbersome, especially for complex types

> But the tests tend to cover corner cases, in a normal script there would
> be far fewer need for type cases.

What is normal, and what is a corner case is subjective anyway.  I hope my config is not full of corner cases (I suspect it is).

**Comment by lacygoill** [comment]
I wonder how TypeScript handles this issue.  BTW, they seem to have a nice alternative syntax with the `as` keyword.
